BAZEL-bash.md

```diff
@@ -0,0 +1,162 @@
+experience, it does seem like rules working on Windows will also work on Linux
+and macOS.
+
+Some of the default paths Bazel creates for you seem to depend on the platform,
+so be sure to always access dependencies through Bazel's `location` mechanism
+rather than rely on th specific directory structure it has created for you on
+your machine.
+
+## Dependencies
+
+Bazel promises hermeticity, but Bash scripts have access to whatever they want.
+As far as I can tell they are not run in any kind of chroot or container (at
+least on macOS), although they are run with a restricted PATH (on Linux) and in
+a specially set-up folder.
+
+As such, in order to get all of your dependencies available, you need to
+specify them in the relevant attributes: `srcs` for files required during
+build, `data` for files required at run time, and `tools` for executables. If
+you're wondering what the distinction is between run time and build time for
+Bash scripts, well, don't ask me.
+
+## `genrule`: substitutions
+
+The code of a `genrule` directive is the `cmd` attribute, which is a multiline
+string embedding a Bash script. Within that script, you need to use `$$` to
+produce a single `$` in the Bash script itself, as `$()` will be expanded by
+Bazel's own string interpolation. You can use the `location` and `locations`
+functions to get at file paths. You should note that the paths these returns is
+relative to the starting `$PWD` of the script, so if your Bash script `cd`s
+somwhere, you have to take that into account.
+
+Note that the script will start running from (virtually) the root of the
+project, regardless of where in the tree your package sits, in that if you
+include local files they will be under your package's path. Only the files you
+have explicitly specified as depencencies will be present.
+
+`srcs` files will generally be placed by Bazel at their current place in the
+workdir, e.g. if you specify `//:VERSION` you will get the `VERSION` file in
+your current directory when the script starts executing, but if you specify
+local targets they will be under the same path as your current `BUILD.bazel`
+file. Try to prefer using `location` entries when possible. I do not know how
+to do that for local files, though.
+
+You can specify local files (that are not Bazel rule results) using the `glob`
+function in the `srcs` attribute, e.g.
+```
+    srcs = glob([
+        "*.c"
+    ])
+```
+This function returns a Bazel array, which can be concatenated with another
+using `+` (typically a literal array to combine the list of local files with a
+list of Bazel targets).
+
+Note that tools and data dependencies will also be given as relative (to the
+script's starting directory) paths through the `location` function.
+
+## Conditionals
+
+Bazel supports simple conditionals in its rules, though I have not found a way
+to use them in string substitutions. As some rules have different names under
```

Comment:
By string substitution you mean `$(location)` expansion? It's syntax is fairly limited. The details are documented [here](https://docs.bazel.build/versions/master/be/make-variables.html#predefined_label_variables).

Reply:
TIL:
> `location`: A synonym for either `execpath` or `rootpath`, depending on the attribute being expanded. This is legacy pre-Starlark behavior and **not recommended** unless you really know what it does for a particular rule. See #2475 for details.
I'm not going to go back to fix it now, but this is definitely something I'll dig into next time I have to touch Bazel.